File: leetcode/17-letter-combinations-of-a-phone-number.rb
# @param {String} digits
# @return {String[]}
def letter_combinations(digits)
  syms = digits.each_char.map {|x| x.to_sym}  
  com syms, 0
end


def com(syms, b)
  if syms.size == 0
    []
  elsif b == syms.size - 1
    MAPPING[syms[b]].each_char.map {|x| x}
  else
    current = MAPPING[syms[b]].each_char.map {|x| x}
    ans = []
    current.each do |c|
      tail = com(syms, b+1).map {|x| c + x}
      ans += tail
    end
    ans
  end
end

MAPPING = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}

File: leetcode/17-letter-combinations-of-a-phone-number.cpp
/*
   Given a digit string, return all possible letter combinations that the number could represent.

   A mapping of digit to letters (just like on the telephone buttons) is given below.



Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].

Note:
Although the above answer is in lexicographical order, your answer could be in any order you want.
*/

class Solution {
    public:
        vector<string> letterCombinations(string digits) {
            if (digits.length() == 0) {
                return vector<string>();
            }
            if (digits.length() == 1) {
                auto s(mapping[digits[0]]);
                vector<string> result;
                for (int i(0); i<s.length(); ++i) {
                    string ss;
                    ss.push_back(s[i]);
                    result.push_back(ss);
                }
                return result;
            }
            auto results(letterCombinations(digits.substr(1)));
            auto s(mapping[digits[0]]);
            vector<string> finalResults;

            for (int i(0); i<s.length(); ++i) {
                string ss;
                ss.push_back(s[i]);
                for (auto &result : results) {
                    finalResults.push_back(ss + result);
                }
            }
            return finalResults;
        }
    private:
        map<char, string> mapping = {{'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"}};
};
